### PR TITLE
Add support for custom checkboxes and radios

### PR DESCRIPTION
**Pull request by eladlevy**
Implemented custom checkboxes and radios mode as discussed in [3047](https://github.com/react-bootstrap/react-bootstrap/issues/3047).
Includes tests and documentation

image

### CODE REVIEW

File: src/FormCheck.js [hunk 1]
```diff
@@ -93,6 +97,7 @@ class FormCheck extends React.Component {
         {...props}
         type={type}
         ref={inputRef}
+        custom={custom}
```
Comment (jquense) [inline src/FormCheck.js]: I think may make sense to put `custom` on the Context object below with controlId, that way when using the individual components you don't need to explicitly pass `custom` to the input and label

File: www/src/pages/components/forms.js [hunk 2]
```diff
@@ -235,6 +237,56 @@ export default withLayout(function FormControlsSection({ data }) {
...
+        completely custom form elements to replace the browser defaults. They’re
+        built on top of semantic and accessible markup, so they’re solid
+        replacements for any default form control.
+      </p>
+      <LinkedHeading h="3" id="forms-custom-checkboxes-and-radios">
+        Checkboxes and radios
+      </LinkedHeading>
+      <p>
+        Each checkbox and radio is wrapped in a <code>{'<div>'}</code> with a
+        sibling <code>FormCheckInput</code> to create our custom control and a{' '}
```
Comment (jquense) [inline www/src/pages/components/forms.js]: I would shorten this all to something like:

> Custom checkbox and radio styles are achieved with a resourceful use of the `:checked` selector and `::after` psuedo elements, but are Structurally similar to the default `FormCheck`.  By default the checked and indeterminate icons use embedded svg icons from Open Iconic...

File: src/FormCheck.js (hunk 1 above)
Comment (eladlevy) [reply src/FormCheck.js]: Great feedback! I wasn't aware of the Context option. This helps to reduce code and makes the customize render option a lot easier to use.

**Comment by taion** [comment]
It seems like this is slightly easier than adding e.g. `<CustomFormCheck.Input>`. That said this seems _almost_ like something `bsPrefix` can handle, though.

**Comment by taion** [comment]
gah i clicked the wrong button

i meant _if_ we had support for `bsPrefix` from `FormCheck` cascading down into `FormCheck.Input`, then had some special-ish (?) logic to also add `custom-checkbox` or whatever if the bsPrefix is `custom-control`.

is that too magical? probably? (though i like having that sort of `bsPrefix` semantics)

in which case i think a `custom` prop is good

**Comment by eladlevy** [comment]
@taion hmm I can try overriding bsPrefix in the render method. I thought about something like:
```javascript
  const activeBsPrefix = custom ? 'custom-control' : bsPrefix;
   <div
          style={style}
          className={classNames(
            className,
            activeBsPrefix,
            custom && `custom-${type}`,
            inline && `${bsPrefix}-inline`,
          )}
    >
```

Is there a better way to override `bsPrefix` ?
It does saves a bit of logic but if a user passes a custom `bsPrefix` it will override his prop and may cause unexpected behaviour.
WDTH?

**Comment by taion** [comment]
no, i was sort of speculating that a "neat" API would be to do something like `bsPrefix="custom-control"`. but i think that's a little too cute.

i'm okay with this API

**Comment by eladlevy** [comment]
So... @jquense  Do you think this PR can be merged?

**Comment by eladlevy** [comment]
Sure. I've rebased latest master on top of my branch so no dirty merge commits will appear.

**Pull request merged**